hoodie-client/src/main/java/com/uber/hoodie/table/HoodieMergeOnReadTable.java

```diff
@@ -94,4 +118,111 @@ public HoodieMergeOnReadTable(HoodieWriteConfig config,
+                        logger.info("Starting to rollback delta commit " + instant);
+
+                        HoodieCommitMetadata commitMetadata = HoodieCommitMetadata
+                                .fromBytes(this.getCommitTimeline().getInstantDetails(new HoodieInstant(true, instant.getAction(), instant.getTimestamp())).get());
+
+                        stats = jsc.parallelize(commitMetadata.getPartitionToWriteStats().keySet().stream().collect(Collectors.toList()))
+                                .map((Function<String, HoodieRollbackStat>) partitionPath -> {
+                                    // read commit file and (either appeßd delete blocks or delete file)
+                                    Map<FileStatus, Boolean> filesToDeletedStatus = new HashMap<>();
+                                    Map<FileStatus, Long> filesToNumBlocksRollback = new HashMap<>();
+
+                                    // we do not know fileIds for inserts (first inserts are parquet files), delete all parquet files for the corresponding failed commit, if present (same as COW)
+                                    filesToDeletedStatus = super.delete(partitionPath, Arrays.asList(commit));
+
+                                    // append rollback blocks for updates
+                                    commitMetadata.getPartitionToWriteStats().get(partitionPath).stream().filter(wStat -> wStat.getPrevCommit() != HoodieWriteStat.NULL_COMMIT).forEach(wStat -> {
+                                        HoodieLogFormat.Writer writer = null;
+                                        try {
+                                            writer = HoodieLogFormat.newWriterBuilder()
+                                                    .onParentPath(new Path(this.getMetaClient().getBasePath(), partitionPath))
+                                                    .withFileId(wStat.getFileId()).overBaseCommit(wStat.getPrevCommit())
+                                                    .withFs(FSUtils.getFs()).withFileExtension(HoodieLogFile.DELTA_EXTENSION).build();
+                                            Long numRollbackBlocks = 0L;
+                                            // generate metadata
+                                            Map<HoodieMetadataType, String> metadata = Maps.newHashMap();
+                                            metadata.put(HoodieMetadataType.INSTANT_TIME, metaClient.getActiveTimeline().lastInstant().get().getTimestamp());
+                                            metadata.put(HoodieMetadataType.TARGET_INSTANT_TIME, commit);
+                                            // if update belongs to an existing log file
+                                            writer.appendBlock(new HoodieCommandBlock(HoodieCommandBlock.HoodieCommandBlockTypeEnum.ROLLBACK_PREVIOUS_BLOCK, metadata));
+                                            numRollbackBlocks++;
+                                            if(wStat.getNumDeletes() > 0) {
```

Comment:
why do we need an additional rollback block for deletes? Its all in the same commit right.. so should nt matter?

Reply:
Please look at `HoodieAppendHandle`, we write 2 different blocks for data blocks and delete blocks.